### PR TITLE
Split nodes to master and data

### LINKED ISSUE #5: Split ES nodes into master, data and client

### PR DESCRIPTION
**Pull request by pavolloffay**
Partially resolves #5 

* split nodes into master (3) and data (2)
* increase resources to 1Gi

Signed-off-by: Pavol Loffay <ploffay@redhat.com>

### CODE REVIEW

**Comment by pavolloffay** [comment]
Master node a couple of times showes:

> 2018-04-12T13:25:50,293][WARN ][i.f.e.d.k.KubernetesUnicastHostsProvider] [SH4mRm0] Exception caught during discovery: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://kubernetes.default.svc/api/v1/namespaces/elasticsearch/endpoints/elasticsearch-cluster. Message: Forbidden!Configured service account doesn't have access. Service account may have been revoked..

But then it outputs 

> [2018-04-12T13:25:59,349][INFO ][o.e.c.s.ClusterService   ] [SH4mRm0] detected_master {5V0IorK}{5V0IorKySa-0te8UfePtrQ}{VRHuX1dFQr-yeadWV1iUsQ}{172.17.0.4}{172.17.0.4:9300}, added {{ZqjSvjP}{ZqjSvjPvRWS65B-O2U5cwA}{oFD7nvBDR3av3l2pn2msOQ}{172.17.0.6}{172.17.0.6:9300},{Nz_HsgH}{Nz_HsgHWT9e2JncXJyPz-g}{3ZJMPscMRQazfFwYq7tKug}{172.17.0.3}{172.17.0.3:9300},{5V0IorK}{5V0IorKySa-0te8UfePtrQ}{VRHuX1dFQr-yeadWV1iUsQ}{172.17.0.4}{172.17.0.4:9300},}, reason: zen-disco-receive(from master [master {5V0IorK}{5V0IorKySa-0te8UfePtrQ}{VRHuX1dFQr-yeadWV1iUsQ}{172.17.0.4}{172.17.0.4:9300} committed version [1]]

I am not sure if it is correct. http://localhost:9200/_cat/nodes whows correct number of nodes in the cluster.

```bash
oc logs po/elasticsearch-master-1759709098-mbmnp 
[2018-04-12T13:25:42,068][INFO ][o.e.n.Node               ] [] initializing ...
[2018-04-12T13:25:42,150][INFO ][o.e.e.NodeEnvironment    ] [SH4mRm0] using [1] data paths, mounts [[/elasticsearch/persistent (/dev/sda1)]], net usable_space [13.6gb], net total_space [17.8gb], spins? [possibly], types [ext4]
[2018-04-12T13:25:42,151][INFO ][o.e.e.NodeEnvironment    ] [SH4mRm0] heap size [494.9mb], compressed ordinary object pointers [true]
[2018-04-12T13:25:42,152][INFO ][o.e.n.Node               ] node name [SH4mRm0] derived from node ID [SH4mRm0oRjyO4Btt0f8RQg]; set [node.name] to override
[2018-04-12T13:25:42,153][INFO ][o.e.n.Node               ] version[5.5.2], pid[1], build[b2f0c09/2017-08-14T12:33:14.154Z], OS[Linux/4.4.41-boot2docker/amd64], JVM[Oracle Corporation/OpenJDK 64-Bit Server VM/1.8.0_161/25.161-b14]
[2018-04-12T13:25:42,153][INFO ][o.e.n.Node               ] JVM arguments [-XX:+UnlockExperimentalVMOptions, -XX:+UseCGroupMemoryLimitForHeap, -XX:MaxRAMFraction=2, -XX:InitialRAMFraction=2, -XX:MinRAMFraction=2, -XX:+UseConcMarkSweepGC, -XX:CMSInitiatingOccupancyFraction=75, -XX:+UseCMSInitiatingOccupancyOnly, -XX:+AlwaysPreTouch, -Xss1m, -Djava.awt.headless=true, -Dfile.encoding=UTF-8, -Djna.nosys=true, -XX:-OmitStackTraceInFastThrow, -Dio.netty.noUnsafe=true, -Dio.netty.noKeySetOptimization=true, -Dio.netty.recycler.maxCapacityPerThread=0, -Dlog4j.shutdownHookEnabled=false, -Dlog4j2.disable.jmx=true, -XX:+HeapDumpOnOutOfMemoryError, -Des.path.home=/usr/share/elasticsearch]
[2018-04-12T13:25:43,332][INFO ][o.e.p.PluginsService     ] [SH4mRm0] loaded module [aggs-matrix-stats]
[2018-04-12T13:25:43,332][INFO ][o.e.p.PluginsService     ] [SH4mRm0] loaded module [ingest-common]
[2018-04-12T13:25:43,332][INFO ][o.e.p.PluginsService     ] [SH4mRm0] loaded module [lang-expression]
[2018-04-12T13:25:43,332][INFO ][o.e.p.PluginsService     ] [SH4mRm0] loaded module [lang-groovy]
[2018-04-12T13:25:43,332][INFO ][o.e.p.PluginsService     ] [SH4mRm0] loaded module [lang-mustache]
[2018-04-12T13:25:43,332][INFO ][o.e.p.PluginsService     ] [SH4mRm0] loaded module [lang-painless]
[2018-04-12T13:25:43,332][INFO ][o.e.p.PluginsService     ] [SH4mRm0] loaded module [parent-join]
[2018-04-12T13:25:43,333][INFO ][o.e.p.PluginsService     ] [SH4mRm0] loaded module [percolator]
[2018-04-12T13:25:43,333][INFO ][o.e.p.PluginsService     ] [SH4mRm0] loaded module [reindex]
[2018-04-12T13:25:43,333][INFO ][o.e.p.PluginsService     ] [SH4mRm0] loaded module [transport-netty3]
[2018-04-12T13:25:43,333][INFO ][o.e.p.PluginsService     ] [SH4mRm0] loaded module [transport-netty4]
[2018-04-12T13:25:43,334][INFO ][o.e.p.PluginsService     ] [SH4mRm0] loaded plugin [discovery-kubernetes]
[2018-04-12T13:25:45,834][INFO ][o.e.d.DiscoveryModule    ] [SH4mRm0] using discovery type [zen]
[2018-04-12T13:25:46,643][INFO ][o.e.n.Node               ] initialized
[2018-04-12T13:25:46,643][INFO ][o.e.n.Node               ] [SH4mRm0] starting ...
[2018-04-12T13:25:46,867][INFO ][o.e.t.TransportService   ] [SH4mRm0] publish_address {172.17.0.5:9300}, bound_addresses {[::]:9300}
[2018-04-12T13:25:46,887][INFO ][o.e.b.BootstrapChecks    ] [SH4mRm0] bound or publishing to a non-loopback or non-link-local address, enforcing bootstrap checks
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[2018-04-12T13:25:47,267][WARN ][i.f.e.d.k.KubernetesUnicastHostsProvider] [SH4mRm0] Exception caught during discovery: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://kubernetes.default.svc/api/v1/namespaces/elasticsearch/endpoints/elasticsearch-cluster. Message: Forbidden!Configured service account doesn't have access. Service account may have been revoked..
[2018-04-12T13:25:50,275][WARN ][o.e.d.z.ZenDiscovery     ] [SH4mRm0] not enough master nodes discovered during pinging (found [[Candidate{node={SH4mRm0}{SH4mRm0oRjyO4Btt0f8RQg}{sJBRpt4ET5aqh_oHLLCW9A}{172.17.0.5}{172.17.0.5:9300}, clusterStateVersion=-1}]], but needed [2]), pinging again
[2018-04-12T13:25:50,293][WARN ][i.f.e.d.k.KubernetesUnicastHostsProvider] [SH4mRm0] Exception caught during discovery: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://kubernetes.default.svc/api/v1/namespaces/elasticsearch/endpoints/elasticsearch-cluster. Message: Forbidden!Configured service account doesn't have access. Service account may have been revoked..
[2018-04-12T13:25:53,297][WARN ][o.e.d.z.ZenDiscovery     ] [SH4mRm0] not enough master nodes discovered during pinging (found [[Candidate{node={SH4mRm0}{SH4mRm0oRjyO4Btt0f8RQg}{sJBRpt4ET5aqh_oHLLCW9A}{172.17.0.5}{172.17.0.5:9300}, clusterStateVersion=-1}]], but needed [2]), pinging again
[2018-04-12T13:25:53,312][WARN ][i.f.e.d.k.KubernetesUnicastHostsProvider] [SH4mRm0] Exception caught during discovery: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://kubernetes.default.svc/api/v1/namespaces/elasticsearch/endpoints/elasticsearch-cluster. Message: Forbidden!Configured service account doesn't have access. Service account may have been revoked..
[2018-04-12T13:25:59,349][INFO ][o.e.c.s.ClusterService   ] [SH4mRm0] detected_master {5V0IorK}{5V0IorKySa-0te8UfePtrQ}{VRHuX1dFQr-yeadWV1iUsQ}{172.17.0.4}{172.17.0.4:9300}, added {{ZqjSvjP}{ZqjSvjPvRWS65B-O2U5cwA}{oFD7nvBDR3av3l2pn2msOQ}{172.17.0.6}{172.17.0.6:9300},{Nz_HsgH}{Nz_HsgHWT9e2JncXJyPz-g}{3ZJMPscMRQazfFwYq7tKug}{172.17.0.3}{172.17.0.3:9300},{5V0IorK}{5V0IorKySa-0te8UfePtrQ}{VRHuX1dFQr-yeadWV1iUsQ}{172.17.0.4}{172.17.0.4:9300},}, reason: zen-disco-receive(from master [master {5V0IorK}{5V0IorKySa-0te8UfePtrQ}{VRHuX1dFQr-yeadWV1iUsQ}{172.17.0.4}{172.17.0.4:9300} committed version [1]])
[2018-04-12T13:25:59,410][INFO ][o.e.h.n.Netty4HttpServerTransport] [SH4mRm0] publish_address {172.17.0.5:9200}, bound_addresses {[::]:9200}
[2018-04-12T13:25:59,410][INFO ][o.e.n.Node               ] [SH4mRm0] started
[2018-04-12T13:26:01,452][INFO ][o.e.c.s.ClusterService   ] [SH4mRm0] added {{0l-gEES}{0l-gEES6S7C0NVBAd-CMRA}{W2slad00SAO7yw0BXzygkg}{172.17.0.7}{172.17.0.7:9300},}, reason: zen-disco-receive(from master [master {5V0IorK}{5V0IorKySa-0te8UfePtrQ}{VRHuX1dFQr-yeadWV1iUsQ}{172.17.0.4}{172.17.0.4:9300} committed version [3]])
```

**Comment by pavolloffay** [comment]
@ruromero could you please review? 
cc @lvlcek

**Comment by pavolloffay** [comment]
@ruromero travis does not seems to start https://github.com/RHsyseng/docker-rhel-elasticsearch/pull/16

**Comment by pavolloffay** [comment]
I have updated the PR to pass the test. 

Locally the deployment created two 
```
NAME                                 DESIRED   CURRENT   READY     AGE
rs/elasticsearch-master-5b77df8676   0         0         0         5m
rs/elasticsearch-master-5ffb79889b    3         3         3         5m
```

describe for 0 0 0 node
```
oc describe rs/elasticsearch-master-5b77df8676               
Name:		elasticsearch-master-5b77df8676
Namespace:	myproject
Selector:	app=elasticsearch,pod-template-hash=1633894232,role=master
Labels:		app=elasticsearch
		pod-template-hash=1633894232
		role=master
Annotations:	deployment.kubernetes.io/desired-replicas=3
		deployment.kubernetes.io/max-replicas=4
		deployment.kubernetes.io/revision=1
		image.openshift.io/triggers=[{
"from": {
  "kind": "ImageStreamTag",
  "name": "elasticsearch:5.5.2"
  },
  "fieldPath": "spec.template.spec.containers[?(@.name==\"elasticsearch-master\")].image"
}]

Replicas:	0 current / 0 desired
Pods Status:	0 Running / 0 Waiting / 0 Succeeded / 0 Failed
Pod Template:
  Labels:	app=elasticsearch
		pod-template-hash=1633894232
		role=master
  Containers:
   elasticsearch-master:
    Image:	 
    Port:	9300/TCP
    Limits:
      memory:	512Mi
    Requests:
      memory:	512Mi
    Liveness:	tcp-socket :cluster delay=60s timeout=1s period=10s #success=1 #failure=3
    Environment:
      SERVICE:		elasticsearch-cluster
      LOG_LEVEL:	info
    Mounts:
      /elasticsearch/persistent from elasticsearch-ephemeral (rw)
      /etc/elasticsearch/elasticsearch.yml from elasticsearch-master-conf (rw)
  Volumes:
   elasticsearch-ephemeral:
    Type:	EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:	
   elasticsearch-master-conf:
    Type:	ConfigMap (a volume populated by a ConfigMap)
    Name:	elasticsearch-master-conf
    Optional:	false
Events:
  FirstSeen	LastSeen	Count	From			SubObjectPath	Type		Reason		Message
  ---------	--------	-----	----			-------------	--------	------		-------
  5m		5m		1	replicaset-controller			Warning		FailedCreate	Error creating: Pod "elasticsearch-master-5b77df8676-tsqlp" is invalid: spec.containers[0].image: Invalid value: " ": must not have leading or trailing whitespace
  5m		5m		1	replicaset-controller			Warning		FailedCreate	Error creating: Pod "elasticsearch-master-5b77df8676-7w494" is invalid: spec.containers[0].image: Invalid value: " ": must not have leading or trailing whitespace
  5m		5m		1	replicaset-controller			Warning		FailedCreate	Error creating: Pod "elasticsearch-master-5b77df8676-7zgwt" is invalid: spec.containers[0].image: Invalid value: " ": must not have leading or trailing whitespace
  5m		5m		1	replicaset-controller			Warning		FailedCreate	Error creating: Pod "elasticsearch-master-5b77df8676-jq67v" is invalid: spec.containers[0].image: Invalid value: " ": must not have leading or trailing whitespace
  5m		5m		1	replicaset-controller			Warning		FailedCreate	Error creating: Pod "elasticsearch-master-5b77df8676-dwn6r" is invalid: spec.containers[0].image: Invalid value: " ": must not have leading or trailing whitespace
  5m		5m		1	replicaset-controller			Warning		FailedCreate	Error creating: Pod "elasticsearch-master-5b77df8676-fx8d4" is invalid: spec.containers[0].image: Invalid value: " ": must not have leading or trailing whitespace
  5m		5m		1	replicaset-controller			Warning		FailedCreate	Error creating: Pod "elasticsearch-master-5b77df8676-gzzld" is invalid: spec.containers[0].image: Invalid value: " ": must not have leading or trailing whitespace
  5m		5m		1	replicaset-controller			Warning		FailedCreate	Error creating: Pod "elasticsearch-master-5b77df8676-pnwfc" is invalid: spec.containers[0].image: Invalid value: " ": must not have leading or trailing whitespace
  5m		5m		1	replicaset-controller			Warning		FailedCreate	Error creating: Pod "elasticsearch-master-5b77df8676-hjdq4" is invalid: spec.containers[0].image: Invalid value: " ": must not have leading or trailing whitespace
  5m		4m		9	replicaset-controller			Warning		FailedCreate	(combined from similar events): Error creating: Pod "elasticsearch-master-5b77df8676-vcgtn" is invalid: spec.containers[0].image: Invalid value: " ": must not have leading or trailing whitespace
```

Could the 000 node fail (because the image entry is empty) multiple times, and then the next one be created?

**Comment by pavolloffay** [comment]
Now it failed on 

> Error from server (AlreadyExists): error when creating "es-cluster-deployment.yml": imagestreams "elasticsearch" already exists

**Comment by pavolloffay** [comment]
The elasticsearch imagestream is created when building/pushing image to the docker registry in prepare script.

**Comment by pavolloffay** [comment]
Locally tests are passing, maybe CI is failing on resources

**Comment by lukas-vlcek** [comment]
Again, I would highly recommend dropping the liveness probe:

```yml
livenessProbe:
  tcpSocket:
    port: cluster
    initialDelaySeconds: 60
```
What do you expect this will buy you? Do you really want to restart the node because it does not response within certain time limit on tcp level? If you want to ping ES nodes then I would go with monitoring tools instead like Prometheus. If you want to make sure you have enough master nodes up and running, again, setup Prometheus rule instead.

If I understand correctly, the way going forward is to have auto-healing implemented via cluster operator that can evaluate specified rules and take action based on that (like restart nodes). We started developing cluster operator for logging ES cluster here https://github.com/openshift/cluster-monitoring-operator/pull/8/files but ATM it does not contain auto-healing part.

**Comment by ruromero** [comment]
I agree with @lukas-vlcek let's remove them

File: es-cluster-deployment.yml [hunk 1]
```diff
@@ -96,6 +207,7 @@ spec:
     targetPort: 9300
   selector:
     app: elasticsearch
+    role: master
```
Comment (ruromero) [inline es-cluster-deployment.yml]: use `elasticsearch: master` or `elasticsearch-role: master`

File: es-cluster-deployment.yml [hunk 2]
```diff
@@ -1,4 +1,111 @@
...
+      "from": {
+        "kind": "ImageStreamTag",
+        "name": "elasticsearch:5.5.2"
+        },
+        "fieldPath": "spec.template.spec.containers[?(@.name==\"elasticsearch-master\")].image"
+      }]
+  name: elasticsearch-master
+  labels:
+    app: elasticsearch
+    role: master
```
Comment (ruromero) [inline es-cluster-deployment.yml]: Use `elasticsearch: master` or `elasticsearch-role: master`

File: es-cluster-deployment.yml [hunk 3]
```diff
@@ -1,4 +1,111 @@
...
+    path:
+      data: /elasticsearch/persistent/elasticsearch/data
+      logs: /elasticsearch/logs
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: elasticsearch-data-conf
+  labels:
+    app: elasticsearch
```
Comment (ruromero) [inline es-cluster-deployment.yml]: Add label as well `elasticsearch: data` or `elasticsearch-role: data`

File: es-cluster-deployment.yml [hunk 4]
```diff
@@ -1,4 +1,111 @@
 ---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: elasticsearch-master-conf
+  labels:
+    app: elasticsearch
```
Comment (ruromero) [inline es-cluster-deployment.yml]: Add label as well `elasticsearch: master` or `elasticsearch-role: master`

File: es-cluster-deployment.yml [hunk 5]
```diff
@@ -1,4 +1,111 @@
...
+    role: master
+spec:
+  replicas: 3
+  template:
+    metadata:
+      labels:
+        app: elasticsearch
+        role: master
```
Comment (ruromero) [inline es-cluster-deployment.yml]: Use `elasticsearch: master` or `elasticsearch-role: master`

**Review by ruromero — CHANGES_REQUESTED** [review]

File: es-cluster-deployment.yml [hunk 6]
```diff
@@ -1,4 +1,111 @@
...
+          value: info
+        resources:
+          limits:
+            memory: 512Mi
+          requests:
+            memory: 512Mi
+        ports:
+        - containerPort: 9300
+          name: cluster
+        livenessProbe:
```
Comment (ruromero) [inline es-cluster-deployment.yml]: Remove probes as agreed

**Comment by pavolloffay** [comment]
TODO:

- [x] move this to a separate file
- [x] remove probes
- [x] fix PR comments

File: es-cluster-deployment.yml (hunk 3 above)
Comment (pavolloffay) [reply es-cluster-deployment.yml]: done

File: es-cluster-deployment.yml (hunk 4 above)
Comment (pavolloffay) [reply es-cluster-deployment.yml]: done

**Comment by pavolloffay** [comment]
Closing see https://github.com/RHsyseng/docker-rhel-elasticsearch/pull/37

**Pull request closed without merging**